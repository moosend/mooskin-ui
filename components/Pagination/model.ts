import { IBoxComponentProps } from '../Box/model';

export interface IPaginationComponentProps extends IBoxComponentProps {
	singleItem?: boolean;

	/** hide show all */
	hideShowAll?: boolean;

	/** number of the active button page */
	activePage: number;

	/** onclick callback when the button is clicked */
	onClickButton?: (e: React.MouseEvent<HTMLElement>, value: number) => void;

	children?: React.ReactElement<IPaginationButtonComponentProps> | Array<React.ReactElement<IPaginationButtonComponentProps>>;
}

export interface IPaginationButtonComponentProps extends IBoxComponentProps {
	/** whether the current button is active */
	active?: boolean;
}
