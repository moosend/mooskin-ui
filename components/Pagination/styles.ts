import styled from 'styled-components';

// Models
import { IBoxComponentProps } from '../Box/model';
import { IPaginationButtonComponentProps, IPaginationComponentProps } from './model';

// Components
import { Box } from '../Box/Box';

import variables from '../_utils/globals/variables';

export const StyledPagination = styled(Box)<IPaginationComponentProps>`
	display: flex;
	align-items: center;
`;

StyledPagination.displayName = 'StyledPagination';

export const StyledPaginationButton = styled(Box)<IPaginationButtonComponentProps>`
	border-radius: 2px;
	font-size: 16px;
	font-weight: 500;
	font-stretch: normal;
	font-style: normal;
	letter-spacing: normal;
	text-align: center;
	cursor: pointer;
	padding: ${(props) => (props.active ? '9px 12px' : '8px 11px')};
	color: ${(props) =>
		props.active
			? `${props.palette?.fontColors.white || variables.fontColors.white}`
			: `${props.palette?.fontColors.medgray1 || variables.fontColors.medgray1}`};
	background-color: ${(props) =>
		props.active
			? `${props.palette?.backgroundColors.primary2 || variables.backgroundColors.primary2}`
			: `${props.palette?.backgroundColors.white || variables.backgroundColors.white}`};
	border: ${(props) => (props.active ? 'none' : `solid 1px ${props.palette?.borderColors.medgray1 || variables.borderColors.medgray1}`)};
	:not(:last-child) {
		margin-right: 10px;
	}
`;

StyledPaginationButton.displayName = 'StyledPaginationButton';

export const StyledPaginationShowAll = styled(Box)<IBoxComponentProps>`
	font-size: 14px;
	font-weight: 500;
	font-stretch: normal;
	font-style: normal;
	letter-spacing: normal;
	text-align: left;
	color: ${(props) => props.palette?.fontColors.primary1 || variables.fontColors.primary1};
	border-bottom: 1px solid ${(props) => props.palette?.fontColors.primary1 || variables.fontColors.primary1};
	cursor: pointer;
`;

StyledPaginationShowAll.displayName = 'StyledPaginationShowAll';
