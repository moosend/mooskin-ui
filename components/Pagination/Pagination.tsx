import * as React from 'react';

import { IconButton } from '../IconButton/IconButton';

// Mooskin Context HoC that passes context to component props
import { withMooskinContext } from '../Styled/MooskinContextProvider';

// Models
import { IPaginationButtonComponentProps, IPaginationComponentProps } from './model';

// Styled Components
import { StyledPagination, StyledPaginationButton, StyledPaginationShowAll } from './styles';

/**
 * Pagination
 */
export const Pagination: React.FC<IPaginationComponentProps> = withMooskinContext((props) => {
	const [showAll, setShowAll] = React.useState(false);
	const [showFirst, setShowFirst] = React.useState(false);
	const [showPrevious, setShowPrevious] = React.useState(false);
	const [showNext, setShowNext] = React.useState(false);
	const [showLast, setShowLast] = React.useState(false);

	React.useEffect(() => {
		setShowFirst(false);
		setShowPrevious(false);
		setShowNext(false);
		setShowLast(false);
	}, []);

	const childrenLength = props.children && Array.isArray(props.children) && props.children.length;

	const batchClickHandler = (e: React.MouseEvent<HTMLElement>, page: number, callback?: (e: React.MouseEvent<HTMLElement>) => void) => {
		props.onClickButton && props.onClickButton(e, page);
		callback && callback(e);
	};

	const onArrowClick = (page: number) => {
		props.onClickButton && props.onClickButton({} as any, page);
	};

	const recurseChildren = (children: any): any => {
		if (!children) {
			return null;
		}

		return React.Children.map(children, (child, i) => {
			if (React.isValidElement<IPaginationButtonComponentProps>(child) && child.type === PaginationButton) {
				const page = i + 1;

				const condition =
					childrenLength && props.activePage >= childrenLength - 1
						? page >= childrenLength - 4
						: props.activePage >= 4
						? page - 2 <= props.activePage && page + 2 >= props.activePage
						: page <= 5;

				if (children[props.activePage - 2] && !showPrevious) {
					setShowPrevious(true);
				} else if (!children[props.activePage - 2] && showPrevious) {
					setShowPrevious(false);
				}

				if (children[props.activePage - 2] && !showFirst) {
					setShowFirst(true);
				} else if (!children[props.activePage - 2] && showFirst) {
					setShowFirst(false);
				}

				if (children[props.activePage] && !showNext) {
					setShowNext(true);
				} else if (!children[props.activePage] && showNext) {
					setShowNext(false);
				}

				if (children[props.activePage] && !showLast) {
					setShowLast(true);
				} else if (!children[props.activePage] && showLast) {
					setShowLast(false);
				}

				if (props.singleItem) {
					return page === props.activePage
						? React.cloneElement(child, {
								active: true,
								children: page,
								key: i
								// onClick: (e) => batchClickHandler(e, page, child.props.onClick),
						  } as IPaginationButtonComponentProps)
						: null;
				}

				if (showAll || condition) {
					return React.cloneElement(child, {
						active: page === props.activePage,
						children: page,
						key: i,
						onClick: (e) => batchClickHandler(e, page, child.props.onClick)
					} as IPaginationButtonComponentProps);
				}
				return null;
			}

			if (React.isValidElement(child) && (child.props as any).children) {
				return React.cloneElement(child, { key: i, children: recurseChildren((child.props as any).children) } as any);
			}

			return child;
		});
	};

	return (
		<StyledPagination {...props}>
			{!showAll && !props.singleItem && (
				<IconButton
					cursor={showFirst ? 'pointer' : 'not-allowed'}
					onClick={showFirst ? () => onArrowClick(1) : undefined}
					fontColor={showFirst ? 'fontColors.text' : 'fontColors.medgray2'}
					opacity={showFirst ? 1 : 0.6}
				>
					first_page
				</IconButton>
			)}
			{!showAll && (
				<IconButton
					cursor={showPrevious ? 'pointer' : 'not-allowed'}
					mr={10}
					onClick={showPrevious ? () => onArrowClick(props.activePage - 1) : undefined}
					fontColor={showPrevious ? 'fontColors.text' : 'fontColors.medgray2'}
					opacity={showPrevious ? 1 : 0.6}
				>
					chevron_left
				</IconButton>
			)}
			{recurseChildren(props.children)}
			{!showAll && (
				<IconButton
					cursor={showNext ? 'pointer' : 'not-allowed'}
					onClick={showNext ? () => onArrowClick(props.activePage + 1) : undefined}
					fontColor={showNext ? 'fontColors.text' : 'fontColors.medgray2'}
					opacity={showNext ? 1 : 0.6}
				>
					chevron_right
				</IconButton>
			)}
			{!showAll && !props.singleItem && (
				<IconButton
					cursor={showLast ? 'pointer' : 'not-allowed'}
					onClick={showLast ? () => childrenLength && onArrowClick(childrenLength) : undefined}
					mr={10}
					fontColor={showLast ? 'fontColors.text' : 'fontColors.medgray2'}
					opacity={showLast ? 1 : 0.6}
				>
					last_page
				</IconButton>
			)}
			{!props.hideShowAll && childrenLength && childrenLength > 5 && (
				<StyledPaginationShowAll palette={props.palette} onClick={() => setShowAll(!showAll)}>
					{showAll ? 'Hide' : 'Show all'}
				</StyledPaginationShowAll>
			)}
		</StyledPagination>
	);
});

Pagination.defaultProps = {
	className: '',
	style: {}
};

Pagination.displayName = 'Pagination';

/**
 * PaginationButton
 */
export const PaginationButton: React.FC<IPaginationButtonComponentProps> = withMooskinContext((props) => {
	return <StyledPaginationButton {...props} />;
});

PaginationButton.defaultProps = {
	className: '',
	style: {}
};

PaginationButton.displayName = 'PaginationButton';
