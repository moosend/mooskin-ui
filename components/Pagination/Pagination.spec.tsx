import * as React from 'react';
import { Pagination, PaginationButton } from './Pagination';

import { mount } from 'enzyme';

describe('Pagination', () => {
	test('renders correctly', () => {
		const func = jest.fn();

		const tree = mount(
			<Pagination onClickButton={func} activePage={3}>
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
			</Pagination>
		);

		expect(tree).toMatchSnapshot();
	});

	test('calls callback correctly', () => {
		const func = jest.fn();

		const tree = mount(
			<Pagination onClickButton={func} activePage={3}>
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
			</Pagination>
		);

		tree.find(PaginationButton).first().simulate('click');
		expect(func).toHaveBeenCalled();
	});

	test('renders pagination button count correctly', () => {
		const func = jest.fn();

		const tree = mount(
			<Pagination onClickButton={func} activePage={3}>
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
				<PaginationButton />
			</Pagination>
		);

		expect(tree.find(PaginationButton).length).toEqual(5);

		tree.find('StyledPaginationShowAll').first().simulate('click');

		expect(tree.find(PaginationButton).length).toEqual(10);
	});
});
