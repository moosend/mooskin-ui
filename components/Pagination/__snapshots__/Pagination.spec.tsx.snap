// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<div
  className="pagination undefined"
>
  <div
    className="paginationInner"
  >
    <PaginationButton
      active={true}
      className="pagination-normal-btn itemButton"
      item={1}
      key="1"
      label="1"
      onClick={[Function]}
    />
    <PaginationButton
      active={false}
      className="pagination-normal-btn itemButton"
      item={2}
      key="2"
      label="2"
      onClick={[Function]}
    />
    <PaginationButton
      active={false}
      className="pagination-normal-btn itemButton"
      item={3}
      key="3"
      label="3"
      onClick={[Function]}
    />
    <PaginationButton
      active={false}
      className="pagination-ellipsis"
      disabled={true}
      item={1}
      key="asdd3"
      label="..."
      onClick={[Function]}
    />
  </div>
</div>
`;
